support/cas-server-support-ehcache3-ticket-registry/src/main/java/org/apereo/cas/config/Ehcache3TicketRegistryConfiguration.java

```diff
@@ -0,0 +1,205 @@
+        } else {
+            val rootDirectory = ehcacheProperties.getRootDirectory();
+            val rootDirectoryFile = new File(rootDirectory);
+            if (!rootDirectoryFile.exists()) {
+                LOGGER.debug("Creating folder for ehcache ticket registry disk cache [{}]", rootDirectory);
+                val mkdirResult = rootDirectoryFile.mkdirs();
+                if (!mkdirResult) {
+                    LOGGER.warn("Unable to create folder for ehcache ticket registry disk cache [{}]", rootDirectory);
+                }
+            }
+            return new DefaultPersistenceConfiguration(rootDirectoryFile);
+        }
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public javax.cache.CacheManager ehcache3TicketCacheManager(
+        @Qualifier ("ehcache3CacheManagerConfiguration") final ServiceCreationConfiguration ehcache3CacheManagerConfiguration) {
+        val ehcacheProperties = casProperties.getTicket().getRegistry().getEhcache3();
+        val cachingProvider = Caching.getCachingProvider(EhcacheCachingProvider.class.getName());
+        val ehcacheProvider = (EhcacheCachingProvider) cachingProvider;
+        val statisticsAllEnabled = ehcacheProperties.isEnableStatistics() ? ConfigurationElementState.ENABLED : ConfigurationElementState.DISABLED;
+        val managementAllAllEnabled = ehcacheProperties.isEnableManagement() ? ConfigurationElementState.ENABLED : ConfigurationElementState.DISABLED;
+        val jsr107Config = new Jsr107Configuration(null, new HashMap<>(),
+            false, managementAllAllEnabled, statisticsAllEnabled);
+        val configuration = new DefaultConfiguration(ehcacheProvider.getDefaultClassLoader(),
+            ehcache3CacheManagerConfiguration,
+            jsr107Config);
+        return ehcacheProvider.getCacheManager(ehcacheProvider.getDefaultURI(), configuration);
+    }
+
+    private CacheConfiguration<String, Ticket> buildCacheConfiguration(final TicketDefinition ticketDefinition) {
+        val ehcacheProperties = casProperties.getTicket().getRegistry().getEhcache3();
+        val terracottaClusterUri = ehcacheProperties.getTerracottaClusterUri();
+
+        CacheEventListenerConfigurationBuilder cacheEventListenerConfiguration = CacheEventListenerConfigurationBuilder
+            .newEventListenerConfiguration(new CasCacheEventListener(),
+                EventType.CREATED, EventType.UPDATED, EventType.EXPIRED, EventType.REMOVED, EventType.EVICTED)
+            .ordered().asynchronous();
+
+        val storageTimeout = ticketDefinition.getProperties().getStorageTimeout();
+        val expiryPolicy = ehcacheProperties.isEternal()
+            ?
+            ExpiryPolicyBuilder.noExpiration()
+            :
+            ExpiryPolicyBuilder.expiry()
+            .create(Duration.ofSeconds(storageTimeout))
+            .access(Duration.ofSeconds(storageTimeout))
+            .update(Duration.ofSeconds(storageTimeout)).build();
```

Comment:
This is interesting. How do we handle idle timeouts here, and specially those that might be different from hard-timeouts? Can this handle an expiration policy to say "2 hours of idleness with a 8-hour hard timeout" ? I suspect you might need to design your own ExpiryPolicy of some kind to handle and separate create timeouts from update/access timeouts.

Reply:
I changed this to be the built-in "timeToIdleExpiration". This post https://stackoverflow.com/questions/43306560/changing-the-ttl-of-individual-entries-in-ehcache-3?rq=1 talks about differences between ehcache2 and 3 with regards to expiration. The ehcache 2 idle timeouts were set to the same thing as the overall timeout and I would think CAS's built-in registry cleaner would kick in to time things out if they are supposed to timeout, is that not the case?

I tried to address all your other feedback and it looks like all tests passed on latest travis run. Let me know if there are more issues.

Reply:
Thank you. I'll review. 

> The ehcache 2 idle timeouts were set to the same thing as the overall timeout and I would think CAS's built-in registry cleaner would kick in to time things out if they are supposed to timeout, is that not the case?

That's correct, yes. I was hoping the registry would be able to clean up after itself removing the need for the cleaner to be turned on. If you set up a policy with an 8-hour hard timeout and a 2-hour idle timeout, it would be easier for the registry to automatically remove/expire the item after 2 hours, rather than having the cleaner pull the registry, calculate expiration periods and remove stuff. The operations to fetch available-tickets from ehcache and remove those that are invalidated by idle timeouts generally tends to cause massive heap surges and spikes, specially over a cluster, unless v3 changes things in that area to improve performance. (i.e. this is how hazelcast works)

It's fine if v3 does the same thing as v2 and we are keeping the same behavior anyway; this is something we can re-iterate and tune later.

Reply:
This all looks great to me; the only thing that I am not sure about is, I think the call here should change to `timeToLiveExpiration()` instead. The TTL is what you'd want. The storage-timeout you get from the catalog is typically the hard timeout of a ticket expiration policy, which should be equal to the lifetime/ttl of an item. Effectively, this should be "CAS says tokens should be kept for 8 hours; ehcache keeps tickets alive for 8 hours"...and then we'd use a cleaner to handle idle timeouts as a separate process. Using `storageTimeout` values for `timeToIdleExpiration()` seems like a mismatch.